log.go

```diff
@@ -177,6 +132,93 @@ func (el *eventLogger) Event(ctx context.Context, event string, metadata ...Logg
 	WriterGroup.Write(append(out, '\n'))
 }
 
+// EventBegin starts an EventInProgress and returns it. The eip is to be
+// completed at a later time
+// Example usage:
+//
+// func SomeFunction() (err error) {
+//    eip := log.EventBegin(ctx, "DoesSomething")
+//    defer func() {
+//        eip.DoneWithErr(err)
+//    }()
+//    ...
+//  }
+func (el *eventLogger) EventBegin(ctx context.Context, event string, metadata ...Loggable) *EventInProgress {
+	_, eip := el.eventBeginHelper(ctx, event, metadata...)
+	return eip
+}
+
+// EventBeginInContext starts an EventInProgress, stores it in the context, and
+// returns the new context. The eip can be completed at a later time using the
+// `MaybeFinishEvent()` method
+// Example usage:
+//
+// func SomeFunction(ctx) {
+//    ctx := log.EventBeginInContext(ctx, "DoesSomething")
+//    ...
+//    go func() {
+//        defer logging.MaybeFinishEvent(ctx)
+//        ...
+//    }
+//    ...
+//  }
+func (el *eventLogger) EventBeginInContext(ctx context.Context, event string, metadata ...Loggable) context.Context {
+	ctx, eip := el.eventBeginHelper(ctx, event, metadata...)
+	return context.WithValue(ctx, activeEventKey, eip)
+}
+
+// MaybeFinishedEvent completes an event associated with ctx.
+// Example usage:
+//
+// func SomeFunction(ctx) {
+//    ctx := log.EventBeginInContext(ctx, "DoesSomething")
+//    ...
+//    go func() {
+//        defer logging.MaybeFinishEvent(ctx)
+//        ...
+//    }
+//    ...
+//  }
+func MaybeFinishEvent(ctx context.Context) {
+	val := ctx.Value(activeEventKey)
+	if eip, ok := val.(*EventInProgress); ok {
+		eip.Done()
+	}
+}
+
+// A helper function for creating events
+func (el *eventLogger) eventBeginHelper(ctx context.Context, event string, metadata ...Loggable) (context.Context, *EventInProgress) {
+	start := time.Now()
+	el.Event(ctx, fmt.Sprintf("%sBegin", event), metadata...)
+
+	span, ctx := opentrace.StartSpanFromContext(ctx, event)
```

Comment:
Do we want to disable this if tracing is disabled?

Reply:
If tracing is disabled then `GlobalTracer` === `NoOpTracer` and operations like this will result in a [noop](https://github.com/opentracing/opentracing-go/blob/master/globaltracer.go#L11), we *do* perform the [check below](https://github.com/ipfs/go-log/pull/18/files/f9a95986416b53e4d9bfd832f26049462f751288#diff-308aaa719425f4333033c97cfb3b9868R204) to avoid looping over the log metadata.

Reply:
Got it. I just want to avoid as much work as possible in that case.